micro_ros_diagnostic_bridge/src/micro_ros_diagnostic_bridge/micro_ros_diagnostic_bridge.cpp

```diff
@@ -0,0 +1,228 @@
+
+#include <rclcpp/parameter.hpp>
+#include <rclcpp/parameter_map.hpp>
+#include <rcl_yaml_param_parser/parser.h>
+
+using micro_ros_diagnostic_msgs::msg::MicroROSDiagnosticStatus;
+using diagnostic_msgs::msg::DiagnosticStatus;
+
+MicroROSDiagnosticBridge::MicroROSDiagnosticBridge(const std::string & path)
+: Node("micro_ros_diagnostic_bridge"),
+  logger_(rclcpp::get_logger("MicroROSDiagnosticBridge"))
+{
+  // Read lookup table
+  declare_parameter("lookup_table", rclcpp::ParameterValue(path));
+  std::string lookup_table_path = get_parameter("lookup_table").as_string();
+  if (lookup_table_path.empty()) {
+    throw std::invalid_argument("Need path to lookup table.");
+  }
+  read_lookup_table(lookup_table_path);
+
+  auto callback =
+    [this](const MicroROSDiagnosticStatus::SharedPtr msg_in) -> void
+    {
+      RCLCPP_DEBUG(
+        get_logger(),
+        "Bridging message from hardware %d, updater %d",
+        msg_in->hardware_id, msg_in->updater_id);
+      msg_out_ = std::make_unique<diagnostic_msgs::msg::DiagnosticStatus>();
+
+      auto updater = lookup_updater(msg_in->updater_id);
+      auto hardware = lookup_hardware(msg_in->hardware_id);
+      msg_out_->hardware_id = hardware;
+      msg_out_->name = updater.first;
+      msg_out_->message = updater.second;
+      msg_out_->level = msg_in->level;
+
+      diagnostic_msgs::msg::KeyValue keyvalue;
+      keyvalue.key = lookup_key(msg_in->updater_id, msg_in->key);
+      switch (msg_in->value_type) {
+        case micro_ros_diagnostic_msgs::msg::MicroROSDiagnosticStatus::VALUE_BOOL:
+          keyvalue.value = std::to_string(msg_in->bool_value);
+          break;
+        case micro_ros_diagnostic_msgs::msg::MicroROSDiagnosticStatus::VALUE_INT:
+          keyvalue.value = std::to_string(msg_in->int_value);
+          break;
+        case micro_ros_diagnostic_msgs::msg::MicroROSDiagnosticStatus::VALUE_DOUBLE:
+          keyvalue.value = std::to_string(msg_in->double_value);
+          break;
+        case micro_ros_diagnostic_msgs::msg::MicroROSDiagnosticStatus::VALUE_LOOKUP:
+          keyvalue.value = lookup_value(msg_in->updater_id, msg_in->key, msg_in->value_id);
+          break;
+      }
+      msg_out_->values.push_back(keyvalue);
+
+      ros2_pub_->publish(std::move(msg_out_));
+    };
+
+  uros_sub_ = create_subscription<MicroROSDiagnosticStatus>(
+    UROS_DIAGNOSTICS_BRIDGE_TOPIC_IN,
+    rclcpp::SystemDefaultsQoS(),
+    callback);
+  ros2_pub_ = create_publisher<DiagnosticStatus>(
+    UROS_DIAGNOSTICS_BRIDGE_TOPIC_OUT,
+    rclcpp::SystemDefaultsQoS());
+}
+
+const std::string
+MicroROSDiagnosticBridge::lookup_key(
+  unsigned int updater_id,
+  unsigned int key)
+{
+  try {
+    return key_map_.at(std::make_pair(updater_id, key));
+  } catch (std::out_of_range & e) {
+    RCLCPP_ERROR(
+      get_logger(),
+      "Updater %d and key %d, not found in lookup table.",
+      updater_id, key);
+    return "NOTFOUND";
+  }
+}
+
+const std::string
```

Comment:
```suggestion
std::string
```
https://stackoverflow.com/questions/8716330/purpose-of-returning-by-const-value